### PR TITLE
[sanitizer] Disable writes to log files for binaries in a secure context.

### PR DESCRIPTION
**Pull request by bigb4ng**
Fix for https://github.com/google/sanitizers/issues/1130.

An original issue described by Szabolcs Nagy at https://seclists.org/oss-sec/2016/q1/363.

Implemented by disabling setting `log_path` in ASAN_OPTIONS to values other then "stderr" and "stdout".

The fix provided for Linux and Android API 18+ based on `AT_SECURE`  auxv variable (`man 3 getauxval` is your friend)..

### CODE REVIEW

File: compiler-rt/lib/sanitizer_common/sanitizer_file.cpp [hunk 1]
```diff
@@ -104,6 +111,16 @@ void ReportFile::SetReportPath(const char *path) {
     }
   }
 
+#ifdef HAS_GETAUXVAL
+  if (getauxval(AT_SECURE) != 0 && path &&
+      internal_strcmp(path, "stderr") != 0 &&
+      internal_strcmp(path, "stdout") != 0) {
+    Report(
+        "ERROR: Permission denied setting log_path for a binary in a secure context. You must run on a same priviledge level\n");
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_file.cpp:119]: nit: typo in "privilege".

Also I know what this is fixing, but I still don't really understand "You must run on a same privilege level". Also I feel talking about permission denied implies something is wrong with the filesystem permissions of the given file.

Maybe something like: "log_path must be 'stderr' or 'stdin' for AT_SECURE (e.g. setuid binaries), is '%s'"

Comment (bigb4ng) [reply compiler-rt/lib/sanitizer_common/sanitizer_file.cpp]: Actually couldn't agree more :)

File: compiler-rt/lib/sanitizer_common/sanitizer_file.cpp [hunk 2]
```diff
@@ -104,6 +111,18 @@ void ReportFile::SetReportPath(const char *path) {
     }
   }
 
+#ifdef HAS_GETAUXVAL
+  if (getauxval(AT_SECURE) != 0 && path &&
+      internal_strcmp(path, "stderr") != 0 &&
+      internal_strcmp(path, "stdout") != 0) {
+    Report(
+        "ERROR: log_path must be 'stderr' or 'stdin' for AT_SECURE "
+        "(e.g. setuid binaries), is '%s\n",
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_file.cpp:120]: nit: missing ' after %s

**Review by fmayer — APPROVED** [review]
Thank you for your contribution.

LGTM, looks like a clear improvement. More generally, I would kind of prefer if each option was either allowed or disallowed with AT_SECURE, but that means we would need a separate one for 'stdout' and 'stderr' only. Let's merge this for now, and think of a more general solution later.

**Comment by fmayer** [comment]
> ⚠️ C/C++ code formatter, clang-format found issues in your code. ⚠️
> 
> You can test this locally with the following command:
> View the diff from clang-format here.

Please run `git clang-format HEAD~`

File: compiler-rt/lib/sanitizer_common/sanitizer_file.cpp [hunk 3]
```diff
@@ -19,6 +19,13 @@
 
 #include "sanitizer_common.h"
 #include "sanitizer_file.h"
+
+#if SANITIZER_LINUX || (SANITIZER_ANDROID && __ANDROID_API__ >= 18)
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_file.cpp:23]: Actually, I am not sure this works. Wouldn't `SANITIZER_LINUX` also be true for Android? I think we can just drop the Android bit, as ASan wasn't supported at API level 18 anyway.

Comment (bigb4ng) [reply compiler-rt/lib/sanitizer_common/sanitizer_file.cpp:23]: You're right, just found out about `sanitizer_getauxval.h`, so I'll use that definition instead. Is `!SANITIZER_GO` necessary though?  https://github.com/bigb4ng/llvm-project/blob/70efd00055d734e8ca10be3c60ca689906f79924/compiler-rt/lib/sanitizer_common/sanitizer_getauxval.h#L26

Comment (fmayer) [reply compiler-rt/lib/sanitizer_common/sanitizer_file.cpp:23]: I'll try to find out if https://github.com/llvm/llvm-project/commit/6c75db8b4bc59eace18143ce086419d37da24746 is still relevant and will let you know as soon as I find out.

Comment (fmayer) [reply compiler-rt/lib/sanitizer_common/sanitizer_file.cpp:23]: But what you can do is use the weak definition from that file

`if (getauxval && getauxval(kAtSecure))`

but you'd have to hardcode the `kAtSecure = 23`

**Comment by fmayer** [comment]
Sorry for the bother, but please do not format parts of the CL that you didn't change. Did you run `git clang-format HEAD~`? That should only format your diff.

**Comment by bigb4ng** [comment]
@fmayer Yeah, my bad. No bother at all was already on it :)

File: compiler-rt/lib/sanitizer_common/sanitizer_file.cpp [hunk 4]
```diff
@@ -104,6 +105,17 @@ void ReportFile::SetReportPath(const char *path) {
     }
   }
 
+  // This is so we can use the weak definition from sanitizer_getauxval.h
+  if (&getauxval && getauxval(/* AT_SECURE */ 23) != 0 && path &&
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_file.cpp:109]: This needs to be under SANITIZER_LINUX, otherwise we don't get the weak definition

Comment (fmayer) [reply compiler-rt/lib/sanitizer_common/sanitizer_file.cpp:109]: Hm, no. sorry. You can't depend on libc in this file.

Comment (fmayer) [reply compiler-rt/lib/sanitizer_common/sanitizer_file.cpp:109]: I think you need to implement what needs the `getauxval` (e.g. `IsAT_SECURE()`) in `sanitizer_libcdep.cpp` (with a stub for non-linux platforms), and put a stub (that returns `false`) into `sanitizer_common_nolibc.cpp`)

File: compiler-rt/lib/sanitizer_common/sanitizer_linux_libcdep.cpp [hunk 5]
```diff
@@ -1071,6 +1071,16 @@ void InitializePlatformCommonFlags(CommonFlags *cf) {
 #  endif
 }
 
+#  if SANITIZER_LINUX
+bool ShouldTreatRuntimeSecurely() {
+  // So we can use the weak definition from sanitizer_getauxval.h
+  if (&getauxval)
+    return getauxval(/* AT_SECURE */ 23) != 0;
+  else
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_linux_libcdep.cpp:1079]: nit: get rid of the else (https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return)

File: compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp [hunk 6]
```diff
@@ -518,6 +518,12 @@ bool IsStateDetached(int state) {
   return state == PTHREAD_CREATE_DETACHED;
 }
 
+#  if !SANITIZER_LINUX
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp:521]: Wouldn't it be easier to put all code here, and make it:

```
bool ShouldTreatRuntimeSecurely() {
#ifdef SANITIZER_LINUX
  if (&getauxval)
    return getauxval(/* AT_SECURE */ 23) != 0;
#endif
  return getuid() == geteuid() && getgid() == getegid();
}
```

File: compiler-rt/lib/sanitizer_common/sanitizer_common_nolibc.cpp [hunk 7]
```diff
@@ -26,6 +26,7 @@ void LogMessageOnPrintf(const char *str) {}
 void WriteToSyslog(const char *buffer) {}
 void Abort() { internal__exit(1); }
 bool CreateDir(const char *pathname) { return false; }
+bool ShouldTreatRuntimeSecurely() { return false; }
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_common_nolibc.cpp:29]: I think this is fine for now, but we should leave a TODO to revist whether this is the correct default. I can see arguments both ways "don't break anything" or "be more secure by default".

**Review by fmayer — APPROVED** [review]

**Comment by fmayer** [comment]
Thanks a lot for your work, looks good to me. Sending to @vitalybuka just to get a second set of eyes on it.

File: compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp [hunk 8]
```diff
@@ -518,6 +522,15 @@ bool IsStateDetached(int state) {
   return state == PTHREAD_CREATE_DETACHED;
 }
 
+bool ShouldTreatRuntimeSecurely() {
+#  if SANITIZER_LINUX
+  // So we can use weak reference from sanitizer_getauxval.h
+  if (&getauxval)
+    return getauxval(/* AT_SECURE */ 23) != 0;
+#  endif
+  return getuid() != geteuid() || getgid() != getegid();
```
Comment (vitalybuka) [inline compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp:531]: why not just return false?

**Review by vitalybuka — COMMENTED** [review]
would be possible to test this?

**Comment by vitalybuka** [comment]
Seems like behavior change which can brake existing users.
Also I don't think it's reasonable to link sanitizer runtimes in security sensitive processes. Sanitizers are bug detection tools, but runtime code was not build as mitigation. Only -fsanitizer-trap= can run without runtimes, but it's supported only UBSAN.

So I am not sure that we should break behavior to improve "not a goal" use cases.

@eugenis

**Comment by fmayer** [comment]
@eugenis suggested this approach in https://github.com/google/sanitizers/issues/1130#issuecomment-524002130

File: compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp (hunk 8 above)
Comment (fmayer) [reply compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp:531]: This is for older versions of glibc that don't have getauxval, or other posix OS that don't have it.

**Comment by vitalybuka** [comment]
> @eugenis suggested this approach in [google/sanitizers#1130 (comment)](https://github.com/google/sanitizers/issues/1130#issuecomment-524002130)

Maybe @eugenis wants to reconsider?

At least I will rerun our internal tests to see if it breaks stuff.

**Comment by bigb4ng** [comment]
@vitalybuka I see your point. As long as there will be a clear warning in the docs about the dangers of using a sanitizer in a production environment, this can become a won't-fix. 

As an alternative, we can consider jailing the path in a smarter way, but there are downsides to this because writing files as a more privileged software albeit in jailed path can become a weaknesses that can potentially be exploited. 

As for testing, it should be possible with some linux magic:
```bash
unshare -Ur
chmod +s ./suid
unshare -U
ASAN_OPTIONS='log_path=a' ./suid
```

**Comment by fmayer** [comment]
@eugenis PTAL

**Comment by eugenis** [comment]
I do not have a strong opinion here. It is certainly useful to build setiud binaries with sanitizers, for testing. It is very convenient to simply build everything with asan/hwasan, without cherry-picking, like we do in Android for memory testing, but with understanding that there are security holes.

We've got ubsan_minimal runtime library that is half way between trap mode and a full sanitizer runtime, developed with security in mind.

If there is no viable plan to make sanitizer runtime production-level secure, then maybe we should err on the side of usability.

**Comment by fmayer** [comment]
> I do not have a strong opinion here. It is certainly useful to build setiud binaries with sanitizers, for testing. It is very convenient to simply build everything with asan/hwasan, without cherry-picking, like we do in Android for memory testing, but with understanding that there are security holes.

I think the question is whether we are likely to need a custom log path for these cases, or not.

**Comment by bigb4ng** [comment]
Upon revisiting this PR, I have devised an alternative solution. We can enhance security by implementing a privilege-dropping mechanism when writing log files. This would involve temporarily dropping effective user and group IDs, as well as effective capabilities, when obtaining a log file descriptor.

This approach should effectively mitigate the attack surface exposed through command-line arguments. However, I have yet to formulate an equivalent solution for addressing potential SELinux use cases. 

I welcome all insights and suggestions.

**Comment by vitalybuka** [comment]
@hctim https://github.com/google/sanitizers/issues/1130 
I'd recommend to close `work as intended`, and just don't do it. I am not convinced it has practical value, but I have limited experience with SELinux.

**Comment by bigb4ng** [comment]
I would strongly recommend against closing an open security issue with publicly available exploits without documenting that such behavior (writing arbitrary files in a context of the running binary) is intended. Otherwise, this might be the best solution after all.

**Comment by hctim** [comment]
I think there's two concerns:
 1. We don't want to break existing users, and
 2. Putting something like this in the code may give the impression that we are really trying very hard to solve this problem, and that it's okay to build setuid binaries with sanitizers and expect it to be a security mitigation.

But it's cheap and solves a known problem.

I'd be in favour of landing a slightly different version of this patch. When we see setuid or some of the other things in AT_SECURE, let's just not do a some cheap stuff (but not die). So:

 1. Disable the symbolizer (not currently in this patch but should happen),
 2. Simply don't allow a log path to be set, don't die, but just return early.
 3. Add a log around both areas saying "you're running a setuid binary under sanitizers, this is not a security mitigation, and you may be actually weakening your security by doing this because the sanitizers were not written with these kind of constraints in mind".

> This approach should effectively mitigate the attack surface exposed through command-line arguments. However, I have yet to formulate an equivalent solution for addressing potential SELinux use cases.

Can you elaborate on what you're concerned about here? A user with selinux permissions to run setuid binaries, and they use sanitizers as a trojan horse to execute syscalls they don't have access to? IMHO this is a "write bad selinux, deal with consequences" problem. I highly doubt any setuid software is designed to not be used as a syscall-trojan-horse.

**Comment by bigb4ng** [comment]
@hctim 
> Can you elaborate on what you're concerned about here? A user with selinux permissions to run setuid binaries, and they use sanitizers as a trojan horse to execute syscalls they don't have access to? IMHO this is a "write bad selinux, deal with consequences" problem. I highly doubt any setuid software is designed to not be used as a syscall-trojan-horse.

Sure! Mostly it's the same concern described in the original writeup. A user with SELinux permissions to run setuid binaries (which is not a "bad selinux" by itself) and a setuid binary compiled with ASAN for "extra security". As there is no warning not to compile setuid with ASAN it may seem harmless and such setuids may end up in certain production environments. 

The issue with SELinux in regard to dropping privilege is that runtime domain is provided based on e.g. file attributes and by design cannot assume privileges of the process owner.

**Comment by fmayer** [comment]
The way `AT_SECURE` interacts with SELinux is as follows: if a binary call results in a domain transition, it is considered `AT_SECURE` because it runs with _different permissions from the caller_.

**Comment by bigb4ng** [comment]
> The way `AT_SECURE` interacts with SELinux is as follows: if a binary call results in a domain transition, it is considered `AT_SECURE` because it runs with _different permissions from the caller_.

Just to clarify, the issue I described is for alternative drop-privilege solution I described. SELinux use-cases will indeed be patched by the current version of the PR.

**Pull request closed without merging**